### PR TITLE
Add `latest-allowed` option from required_version

### LINKED ISSUE #307: 'latest-allowed' version as opposite to 'min-required'
`min-required` is convenient to use the already required `terraform { required_version = ... }`. More often than the oldest version allowed by that spec though, I'd like to be using the latest.

It would be great if (in addition to the `latest` and `latest:<regex>` already supported) tfenv supported a `latest-allowed` (or similar) that would choose the latest version subject to the `required_version`.

### PR DESCRIPTION
**Pull request by OJFord**
This commit adds an option similar to `min-required` in that it parses
the terraform's `required_version`, but then uses the most recent
version allowed by that spec.

For example, given:

    required_version = "~> 0.10.0"

`min-required` would give us `0.10.0`, while `latest-allowed` will give
us `0.10.8` (via `latest:^0.10`).

The `<` operator is not implemented (only `>` `>=` `~>` `<=`) because
it's messy to find the latest version smaller than a given one in
`list-remote`'s output, given it may not (yet) exist, and it seems a
reasonably assumption/requirement that the version numbers will exist,
since, for example:

    required_version = "~> 1.0"

expresses the same as:

    required_version = "< 2.0.0"

without requiring a version that doesn't (at time of writing) exist.

This assumption/requirement stands for `<=` too, but again I think
that's reasonable since `<= 2.0.0` (as opposed to not equal to) seems
like a strange version constraint to use before it exists. (And when it
does, one can just use `2.0.0`.)

Closes #307.

### CODE REVIEW

**Comment by OJFord** [comment]
Actually, I just saw there was a release a few days ago. Note the comments there:
https://github.com/tfutils/tfenv/releases/tag/v2.2.3

**Comment by smelchior** [comment]
@OJFord we would also really like this feature! May i ask you to rebase the branch so merging is easier for the maintainers :)?

**Comment by OJFord** [comment]
@smelchior Done, there weren't actually any conflicts, GitHub's warning was just that it was 'out of date', i.e. there was a commit ('update changelog') on master not on this branch.

I don't think that's what was holding this back though, I wouldn't hold your breath for any merges! Let me know if you run tfenv from this branch / with this patch and run into any issues in the meantime.

@Zordrak Obviously I'm biased by my own PRs, but those aside if you need any help maintaining tfenv, and issue triage etc. I'd be happy to help.

**Comment by OJFord** [comment]
Hi! Good to hear from you. Sure, I'll look into it.

File: README.md
Comment (Zordrak) [inline README.md:91]: It's not recursive, and certainly shouldn't be as terraform does not recurse when parsing code.

File: libexec/tfenv-resolve-version
Comment (Zordrak) [inline libexec/tfenv-resolve-version:145]: Does this case not need to be accounted for in tfenv-version-name and tfenv-uninstall also?

Comment (OJFord) [reply libexec/tfenv-resolve-version:130]: Thanks

File: libexec/tfenv-resolve-version
Comment (OJFord) [inline libexec/tfenv-resolve-version:144]: Seems we can't use a negative count to (BSD) `head` on macOS.

I saw `tfenv` already depends on GNU `grep` on macOS (ed6f2a13cc97ed2b5db11000f1a9b3021be4685e) - do you want to add the `head` (`brew install coreutils`) dependency as well, or shall I find another way to do this?

Comment (jack-parsons-bjss) [reply libexec/tfenv-resolve-version:144]: You can get the first line with `sed` which is already a dependency: `sed -n 1p`

Comment (OJFord) [reply libexec/tfenv-resolve-version:144]: Here I don't want (only) the first line, but the first `N-1` lines.

e.g. from `1.2.3` we want `1.2`; from `1.2`, `1`.

Of course there's other ways of doing it (perhaps not as concisely?) I just wondered whether to bother.

Comment (jack-parsons-bjss) [reply libexec/tfenv-resolve-version:144]: Whoops! Totally misread what you were doing :) It's possible with sed or awk and effort has gone into not adding coreutils as a dependency, so I think it's worth doing.

**Comment by smelchior** [comment]
I would still really like to see this PR merged, is there something still open I could assist with :)?

**Comment by OJFord** [comment]
Thanks @smelchior, so would I (and I'm still here and willing to work on this too fwiw 🙂).

I think I added a passing test as was requested, so good to go from my perspective. 

If you could (if you're not already) use this branch to test it out manually though I'm sure any feedback (whether you have issues or not)  would be appreciated.

**Comment by smelchior** [comment]
Thanks @OJFord i already did some tests on my side and it works as advertised :)
Very nice feature which reduces the hassle of updating the version in 2 places when there is an update!

**Review by Zordrak — APPROVED** [review]

**Comment by smelchior** [comment]
Thanks for merging this @Zordrak, that will really help our workflow :)
Do you have an estimate when a new release will be made including this feature?

**Comment by mputilin** [comment]
@Zordrak please release it 🙏

**Comment by jlestrada** [comment]
@Zordrak any timeline to release this change along with the other commits that were merged in? 

Seems like the community (including myself) would be happy to help as well with anything to get this distributed so let us know what we can help with.

**Comment by timo-reymann** [comment]
Do you have any news when this feature will be released, @Zordrak? :)

**Pull request merged**
